### ISSUE TITLE
Add a user specified timeout parameter to TCP sockets

### ISSUE THREAD

**Issue by shamblett** [shamblett/mqtt_client#245]
TCP sockets allow a timeout value to be passed on connect -
```
Socket.connect(server, port, timeout: new Duration(seconds: 5)).then((dynamic socket) 
```
This could be added as a parameter to the server client.

**Comment by jonny7737** [comment]
I converted from mqtt5 to mqtt to reduce the 60 second timeout issue only to find that the issue exists here as well.  This is the timeout i referred to in my mqtt5 issue report.  Any update on this issue.

**Comment by jonny7737** [comment]
I modified mqtt_client_mqtt_server_normal_connection: connect override: add this code at the top of the Socket.connect().then(){} section

```
        final enableKeepaliveOption = RawSocketOption.fromBool(
            0xffff, // SOL_SOCKET
            0x0008, // SO_KEEPALIVE
            true
        );
        final keepaliveIntervalOption = RawSocketOption.fromInt(
            6, // IPPROTO_TCP
            0x10, // TCP_KEEPALIVE
            2
        );
        final keepaliveSuccessiveIntervalOption = RawSocketOption.fromInt(
          6, // IPPROTO_TCP
          0x101, // TCP_KEEPINTVL
          1,
        );
        final keepaliveusccessiveCountOption = RawSocketOption.fromInt(
            6, // IPPROTO_TCP
            0x102, // TCP_KEEPCNT
            2
        );
        socket.setRawOption(enableKeepaliveOption);
        socket.setRawOption(keepaliveIntervalOption);
        socket.setRawOption(keepaliveSuccessiveIntervalOption);
        socket.setRawOption(keepaliveusccessiveCountOption);

```

FOR MACOS: this code addition provides the expected timeout when the server abruptly disconnects from the network.
This solution is not mine and I have no details on the meaning of the values used to create the RawSocketOption objects.

**Comment by shamblett** [comment]
Hmm, a few unit test are now failing with 'OS Error: Protocol not available, errno = 92', the mqtt_client_connection_unsecure_test.dart is an example, I'll dig deeper into whats happening here.

**Comment by jonny7737** [comment]
In what environment was the failed test run?  I see no such failures when running my app on macos.

**Comment by shamblett** [comment]
It fails on my main linux dev box which runs Fedora 37 as it happens,

**Comment by jonny7737** [comment]
My debian VM shows these values for the TCP_keys:
```
MAX_TCP_KEEPIDLE: 0x7FFF
MAX_TCP_KEEPINTVL: 0x7FFF
MAX_TCP_KEEPCNT: 0x7F
```
Using these or whatever your dev box has in tcp.h.  I would like to know if they are the same as these values.

**Comment by jonny7737** [comment]
My DEbian VM defins the TCP_keys as:

```
#define TCP_KEEPIDLE		4	/* Start keeplives after this period */
#define TCP_KEEPINTVL		5	/* Interval between keepalives */
#define TCP_KEEPCNT		6	/* Number of keepalives before death */

```

Hope this is still helping.

**Comment by shamblett** [comment]
Just quickly commenting out the setting of the options and seeing what happens has revealed that setting any of the options causes the problem, even if you substitute the values from Linux then you would also need values for windows and arm and ultimately RISCV. This will get to messy and will need platform specific code which I don't want to add.

So, I think the best thing to do is to put a generic socket option API on the client, so the user can set up the values they want using addBool, addInt etc. and the client will apply these on connect. This way if it breaks it can easily be undone, the user is free to set any socket option they like on whatever platform they like. 

What do you think?

**Comment by shamblett** [comment]
OK, Ill create a branch for this and add it to the client, shouldn't take too long.

**Comment by shamblett** [comment]
OK, I've updated the server client to accept a list of user supplied RawSocketOptions, if you could point your pubspec at the mqtt_client repo and set the branch to issue245 and look at the ' socketOptions' API it would be good to get some feeback from you before I re publish the package, thanks.

**Comment by jonny7737** [comment]
Simple, easy and works as expected.  Thank you so much.

```
    final enableKeepaliveOption = RawSocketOption.fromBool(0xffff, 0x0008, true);
    final keepaliveIntervalOption = RawSocketOption.fromInt(6, 0x10, 2);
    final keepaliveSuccessiveIntervalOption = RawSocketOption.fromInt(6, 0x101, 1);
    final keepaliveSuccessiveCountOption = RawSocketOption.fromInt(6, 0x102, 1);

    client.socketOptions = [
      enableKeepaliveOption,
      keepaliveIntervalOption,
      keepaliveSuccessiveIntervalOption,
      keepaliveSuccessiveCountOption
    ];

```

**Comment by jonny7737** [comment]
1 out of approx. 50 tests of broker abruptly terminating showed an unhandled exception. Not a problem i think but wanted to let you know just in case:

```
[ERROR:flutter/runtime/dart_vm_initializer.cc(41)] Unhandled Exception: SocketException: Write failed (OS Error: Connection reset by peer, errno = 54), address = 10.211.55.8, port = 53154

#0      _NativeSocket.write (dart:io-patch/socket_patch.dart:1201:34)
#1      _RawSocket.write (dart:io-patch/socket_patch.dart:1944:15)
#2      _Socket._write (dart:io-patch/socket_patch.dart:2387:18)
#3      _SocketStreamConsumer.write (dart:io-patch/socket_patch.dart:2134:26)
#4      _SocketStreamConsumer.addStream.<anonymous closure> (dart:io-patch/socket_patch.dart:2108:11)
#5      _RootZone.runUnaryGuarded (dart:async/zone.dart:1593:10)
#6      _BufferingStreamSubscription._sendData (dart:async/stream_impl.dart:339:11)
#7      _BufferingStreamSubscription._add (dart:async/stream_impl.dart:271:7)
#8      _SyncStreamControllerDispatch._sendData (dart:async/stream_controller.dart:774:19)
#9      _StreamController._add (dart:async/stream_controller.dart:648:7)
#10     _StreamController.add (dart:async/stream_controller.dart:596:5)
#11     _StreamSinkImpl.add (dart:io/io_sink.dart:136:17)
#12     _Socket.add (dart:io-patch/socket_patch.dart:2233:38)
#13     MqttServerNormalConnection.send (package:mqtt_client/src/connectionhandling/server/mqtt_client_mqtt_server_normal_connection.dart:104:13)
#14     MqttConnectionHandlerBase.sendMessage (package:mqtt_client/src/connectionhandling/mqtt_client_mqtt_connection_handler_base.dart:167:18)
#15     MqttConnectionKeepAlive.pingRequired (package:mqtt_client/src/connectionhandling/mqtt_client_mqtt_connection_keep_alive.dart:89:28)
#16     Timer._createTimer.<anonymous closure> (dart:async-patch/timer_patch.dart:18:15)
#17     _Timer._runTimers (dart:isolate-patch/timer_impl.dart:398:19)
#18     _Timer._handleMessage (dart:isolate-patch/timer_impl.dart:429:5)
#19     _RawReceivePort._handleMessage (dart:isolate-patch/isolate_patch.dart:192:26)

```

**Comment by shamblett** [comment]
In your use case I think this is acceptable but thanks for the research anyway.

I've merged the change and re published the package at version 9.8.0, also I've raised this [issue](https://github.com/shamblett/mqtt5_client/issues/53) on the mqtt5_client to update it also.

Closing this.
